integration/gateway/gateway_test.go

```diff
@@ -0,0 +1,120 @@
+	var (
+		testDir       string
+		client        *docker.Client
+		network       *nwo.Network
+		process       ifrit.Process
+		orderer       *nwo.Orderer
+		org1Peer0     *nwo.Peer
+		org2Peer0     *nwo.Peer
+		gatewayClient proto.GatewayClient
+	)
+
+	BeforeEach(func() {
+		var err error
+		testDir, err = ioutil.TempDir("", "gateway")
+		Expect(err).NotTo(HaveOccurred())
+
+		client, err = docker.NewClientFromEnv()
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	AfterEach(func() {
+		if process != nil {
+			process.Signal(syscall.SIGTERM)
+			Eventually(process.Wait(), network.EventuallyTimeout).Should(Receive())
+		}
+		if network != nil {
+			network.Cleanup()
+		}
+		os.RemoveAll(testDir)
+	})
+
+	Describe("evaluate with result", func() {
+		BeforeEach(func() {
+			config := nwo.BasicEtcdRaft()
+			network = nwo.New(config, testDir, client, StartPort(), components)
+
+			network.GatewayEnabled = true
+
+			network.GenerateConfigTree()
+			network.Bootstrap()
+
+			networkRunner := network.NetworkGroupRunner()
+			process = ifrit.Invoke(networkRunner)
+			Eventually(process.Ready(), network.EventuallyTimeout).Should(BeClosed())
+
+			orderer = network.Orderer("orderer")
+			org1Peer0 = network.Peer("Org1", "peer0")
+			org2Peer0 = network.Peer("Org2", "peer0")
+
+			network.CreateAndJoinChannel(orderer, "testchannel")
+			network.UpdateChannelAnchors(orderer, "testchannel")
+			network.VerifyMembership(network.PeersWithChannel("testchannel"), "testchannel")
+
+			nwo.EnableCapabilities(network, "testchannel", "Application", "V2_0", orderer, org1Peer0, org2Peer0)
+
+			chaincodePath, err := filepath.Abs("../chaincode/gateway")
+			Expect(err).NotTo(HaveOccurred())
+
+			chaincode := nwo.Chaincode{
+				Name:            "gatewaycc",
+				Version:         "0.0",
+				Path:            chaincodePath,
+				Lang:            "golang",
+				PackageFile:     filepath.Join(testDir, "gatewaycc.tar.gz"),
+				Ctor:            `{"Args":[]}`,
+				SignaturePolicy: `AND ('Org1MSP.peer')`,
+				Sequence:        "1",
+				InitRequired:    false,
+				Label:           "gatewaycc",
+			}
+
+			nwo.DeployChaincode(network, "testchannel", orderer, chaincode)
+
+			gatewayClient = GatewayClient(
+				network.PeerAddress(org1Peer0, nwo.ListenPort),
+				filepath.Join(network.PeerLocalTLSDir(org1Peer0), "ca.crt"),
+			)
+		})
+
+		Context("when I evaluate an echo transaction with the arguments [\"conga\"]", func() {
+			It("should respond with \"conga\"", func() {
+				ctx, cancel := context.WithTimeout(context.Background(), 100*time.Second)
+				defer cancel()
+
+				signingIdentity := GetSigningIdentity(network, "Org1", "User1", "Org1MSP", "bccsp")
+				txn := NewProposedTransaction(signingIdentity, "testchannel", "gatewaycc", "Echo", []byte("conga"))
+
+				result, err := gatewayClient.Evaluate(ctx, txn)
+				Expect(err).NotTo(HaveOccurred())
+				Expect(result).NotTo(BeNil())
+				Expect(result.Value).To(Equal([]byte("conga")))
```

Comment:
using `proto.Equal` with an expected struct would be preferred.

Reply:
Why is that the preferred option for such a simple struct? It seems to provide less useful output when the test fails:

```
Expected
  <bool>: false
to be true
```

Reply:
When fields are added to a struct, we'd like the assertions to include them. When we assert on individual fields, we're not asserting the compatible zero values of new fields or the non-zero fields in responses that should get tested.

As for proto.Equal, if you go through proto marshaling, depending the version of the proto package, a number of fields are populated that impact the basic equality tests.

If you need a matcher, we can add one. They're trivial:
```
// Copyright IBM Corp. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

package matcher

import (
	"fmt"

	"github.com/onsi/gomega/types"
	"google.golang.org/protobuf/proto"
)

// ProtoEqual is a gomega matcher for protobuf messages.
func ProtoEqual(expected proto.Message) types.GomegaMatcher {
	return &protoEqualMatcher{
		expected: expected,
	}
}

type protoEqualMatcher struct {
	expected proto.Message
}

func (e *protoEqualMatcher) Match(actual interface{}) (bool, error) {
	actualMessage, ok := actual.(proto.Message)
	if !ok {
		return false, fmt.Errorf("ProtoEqual expects a proto.Message")
	}

	return proto.Equal(e.expected, actualMessage), nil
}

func (e *protoEqualMatcher) FailureMessage(actual interface{}) string {
	return fmt.Sprintf("Expected\n\t%#v\nto proto.Equal\n\t%#v", actual, e.expected)
}

func (e *protoEqualMatcher) NegatedFailureMessage(actual interface{}) string {
	return fmt.Sprintf("Expected\n\t%#v\nnot to proto.Equal\n\t%#v", actual, e.expected)
}
```

Reply:
A matcher sounds like it would be nice to have in the future https://jira.hyperledger.org/browse/FAB-10301

Reply:
If it's nice to have, it would be even nicer for someone that needs it to pick up the story and implement it. That would be really nice...

As for the "trivial" nature, I disagree. When behavior changes at this level (the outside most API), it should most certainly require some explicit action. I've tried to explain the rationale for why evaluating the entire message is preferred over a subset.